engine/overworld/weather.asm

```diff
@@ -0,0 +1,778 @@
+DoOverworldWeather:
+	push de
+	push hl
+	push bc
+	ld hl, wWeatherFlags
+	bit OW_WEATHER_DISABLED_F, [hl]
+	jr nz, .done
+	ld a, [wOverworldWeatherDelay]
+	and %1 ; 30 fps
+	jr z, .done
+	ld a, [wOverworldWeatherCooldown]
+	and a
+	jr z, .no_cooldown
+	dec a
+	ld [wOverworldWeatherCooldown], a
+.no_cooldown
+	ldh a, [hUsedOAMIndex]
+	cpl
+	add (SPRITEOAMSTRUCT_LENGTH * NUM_SPRITE_OAM_STRUCTS) + 1
+	ld hl, hUsedWeatherSpriteIndex
+	cp [hl]
+	jr nc, .ok
+	add -SPRITEOAMSTRUCT_LENGTH
+	ldh [hUsedWeatherSpriteIndex], a
+.ok
+	ld a, [wOverworldWeatherCooldown]
+	and a
+	jr nz, .on_cooldown
+	ld a, [wCurrentWeather]
+	dec a ; OW_WEATHER_RAIN
+	call z, DoOverworldRain
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_SNOW
+	call z, DoOverworldSnow
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_THUNDERSTORM
+	call z, DoOverworldRain
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_SANDSTORM
+	call z, DoOverworldSandstorm
+.done
+	ld hl, wOverworldWeatherDelay
+	inc [hl]
+	call WeatherSpriteLimitCheck
+	pop bc
+	pop hl
+	pop de
+	ret
+
+.on_cooldown
+	ld a, [wPrevWeather]
+	cp OW_WEATHER_RAIN
```

Comment:
This series could use the `dec a` series optimization (with appropriate `assert`s).

Reply:
Or maybe a jumptable as well?